File: UTILS/python/add_picks2weight.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

"""
This code takes in a file name with P arrivals and makes a different weight file with the P arrivals.  The new weight file can be a body weight file, surface wieight file or using everything.
Kyle Smith
Made January 19, 2018 
"""

import os, sys
import obspy
from read_cap_weight import *
import math
import string

def add_picks2weight(file_w_Parrivals,outfile,methodnumber,scalewts):
    if 1 == 0:
        methodnumber = 2
        scalewts = 100 
        # name of file that contains the P arrivals
        file_w_Parrivals = "/home/ksmith/REPOSITORIES/capuaf/MTs/Dec10MT/V1/20171210122855089/weight_body_150.dat"

        # name of the output file
        outfile = "/home/ksmith/REPOSITORIES/capuaf/MTs/Dec10MT/V1/20171210122855089/weight_body_150_sc.dat"

    # read in weight file
    stnm,edist,PV_wt,PR_wt,SV_wt,SR_wt,ST_wt,P_arrival,P_len,S_arrival,S_len,waveform_shft = read_cap_weight(file_w_Parrivals)

    # make new file
    f = open(outfile, "w")
    stfmt = '%34s %4.0f %4.0f %4.0f %4.0f %4.0f %4.0f %4.4f %4.0f %4.0f %4.0f %4.0f \n';

    # make copy
    if methodnumber == 1:
        for linenum in range(0,len(stnm)):
            xx = stfmt % (stnm[linenum],edist[linenum],PV_wt[linenum],PR_wt[linenum],SV_wt[linenum],SR_wt[linenum],ST_wt[linenum],P_arrival[linenum],P_len[linenum], S_arrival[linenum],S_len[linenum],waveform_shft[linenum])
            f.write(xx)

    # scale file
    if methodnumber == 2:
        for linenum in range(0,len(stnm)):
            xx = stfmt % (stnm[linenum],edist[linenum],PV_wt[linenum]*scalewts,PR_wt[linenum]*scalewts,SV_wt[linenum]*scalewts,SR_wt[linenum]*scalewts,ST_wt[linenum]*scalewts,P_arrival[linenum],P_len[linenum], S_arrival[linenum],S_len[linenum],waveform_shft[linenum])
            f.write(xx)

    # make standard weight file (all 1s)
    if methodnumber == 3:
        for linenum in range(0,len(stnm)):
            xx = stfmt % (stnm[linenum],edist[linenum],scalewts,scalewts,scalewts,scalewts,scalewts,P_arrival[linenum],P_len[linenum], S_arrival[linenum],S_len[linenum],waveform_shft[linenum])
            f.write(xx)

    # make body waves weight file
    if methodnumber == 4:
        for linenum in range(0,len(stnm)):
            xx = stfmt % (stnm[linenum],edist[linenum],scalewts,scalewts,0,0,0,P_arrival[linenum],P_len[linenum], S_arrival[linenum],S_len[linenum],waveform_shft[linenum])
            f.write(xx)

    # make surface waves weight file
    if methodnumber == 5:
        for linenum in range(0,len(stnm)):
            xx = stfmt % (stnm[linenum],edist[linenum],0,0,scalewts,scalewts,scalewts,P_arrival[linenum],P_len[linenum], S_arrival[linenum],S_len[linenum],waveform_shft[linenum])
            f.write(xx)

    f.close()

    return outfile

File: UTILS/matlab/add_picks2weight.m
function [] = add_picks2weight(weight_input,weight_output,idb,iUseReviewed,HWIN_SEC)
% Adds first-motion polarity and P arrival times to the weight file

%idb = 1;            % Use AEC db
iPonly = 1;         % Get only P arrival time
%HWIN_SEC = 2;       % this is really important because otime is different for AEC and IRIS
%iUseReviewed = 0;   % Analyst reviewed or unreviewed polarity
spdy = 86400;

% Read the weight file
weight_input
if ~exist(weight_input,'file'), error('file does not exist'); end
[stn_tag, pol, edist, PV_wt, PR_wt, SV_wt, SR_wt, ST_wt, P_arrival, P_len, S_arrival, S_len, waveform_shft] = read_cap_weight(weight_input);

% Get the origin time and add the time range (HWIN_SEC) to match with the AEC otime
tmp = strsplit(stn_tag{1},'.');
eid = tmp{1};
otime = eid2otime(eid);
otar = [otime HWIN_SEC];

% Get polarity and time picks
[sta_aec,chan,timepick,iphase,fm,ifm,delta_deg,rlon,rlat,relev,elon,elat,edep,eml,originTime,dist_hyp] = get_picks(otar,idb,iPonly,iUseReviewed);

% KEY: this time shift must be relative to the target origin time, which
% might differ from the origin time that is in the database; the reason is
% that our sac files have the origin time set as otar
p_time_aec = (timepick-originTime)*spdy;

% Difference between AEC origin time and IRIS origin time
odiff = (originTime-otime)*spdy;

% add the difference between otimes' to the P arrival time w.r.t.  AEC
% origin time
p_time_iris = p_time_aec + odiff;

% Match station name and add polarity (ifm) and P arrival time (timepick)
for ii = 1:length(stn_tag)
    tmp = strsplit(stn_tag{ii},'.');
    stnm_iris(ii) = tmp(3);
    for jj = 1:length(sta_aec)
        if strcmp(stnm_iris{ii},sta_aec{jj})
            pol_aec(ii) = ifm(jj);
            P_arrival_aec(ii) = p_time_iris(jj);
        end
    end
end

% write the weight file
write_cap_weight(weight_output, stn_tag, pol_aec, edist, PV_wt,PR_wt, SV_wt, SR_wt, ST_wt, P_arrival_aec, P_len, S_arrival, S_len, waveform_shft)

% Examples

if 0
    weight_input = '/home/vipul/REPOSITORIES/GEOTOOLS/python_util/util_data_syn/20090407201253480/weight.dat';
    weight_output = '/home/vipul/REPOSITORIES/GEOTOOLS/python_util/util_data_syn/20090407201253480/weight_test.dat';
    idb = 1;            % Use AEC db
    HWIN_SEC = 2;       % this is really important because otime is different for AEC and IRIS
    iUseReviewed = 0;   % Analyst reviewed or unreviewed polarity
    add_picks2weight(weight_input,weight_output,idb,iUseReviewed,HWIN_SEC)
    %-----------------------------------------------------------------
%     weight_input = '/home/ksmith/REPOSITORIES/capuaf/Nov20MT/20151120105348168/weight_body.dat';
%     weight_output = '/home/ksmith/REPOSITORIES/capuaf/Nov20MT/20151120105348168/weight_body_picks2.dat';
%     idb = 1;            % Use AEC db
%     HWIN_SEC = 2;       % this is really important because otime is different for AEC and IRIS
%     iUseReviewed = 1;   % Analyst reviewed or unreviewed polarity
%     add_picks2weight(weight_input,weight_output,idb,iUseReviewed,HWIN_SEC)
end
